Isolate Hardhat Network Explanation & add a Reference (#1657)
nomiclabs/hardhat: docs/hardhat-network/reference/README.md

# Hardhat Network Reference

## Supported hardforks

- byzantium
- constantinople
- petersburg
- istanbul

> - muirGlacier

Reviewer: ```suggestion
- muirGlacier
- london
```

## Config

- Default behavior: Hardhat will automatically run in-process when `defaultNetwork` is empty or set to `hardhat`.

> ### Mining modes

Reviewer: Is this intended to bu a subsection of Config?

Reviewer: Well, this section is about config. I think what confuses me are the following ones, that mostly use the RPC

Author: @alcuadrado and I had a call about this, and I took these notes on the intended outcome:
```
medium term, give each product its own PAGE, with its own sidebar.
for short term, re-word "Explanation" to "What is it?", add another page to explain mining modes, and link to it in the sidebar as "Mining Modes", and continue the convention of referring to the Divio content types in a markdown comment in the top-level heading of the page, as already done in the hardhat-network/README.md document.
```
I implemented the short term approach here: https://github.com/nomiclabs/hardhat/pull/1657/commits/3383747fecd278d79aa571eb2f06d7fa63997b98

    }
  }
}
```

In this example, automining is disabled and interval mining is set so that a new block is generated every 5 seconds. You can also configure interval mining to generate a new block after a random delay:

```js
networks: {
  hardhat: {
    mining: {
      auto: false,
      interval: [3000, 6000]
    }
  }
}
```

In this case, a new block will be mined after a random delay of between 3 and 6 seconds. For example, the first block could be mined after 4 seconds, the second block 5.5 seconds after that, and so on.

#### Manual mining

You can disable both mining modes like this:

```js
networks: {
  hardhat: {
    mining: {
      auto: false,
      interval: 0
    }
  }
}
```

This means that no new blocks will be mined by the Hardhat Network, but you can manually mine new blocks using the `evm_mine` RPC method. This will generate a new block that will include as many pending transactions as possible.

### Mempool behavior

When automine is disabled, every sent transaction is added to the mempool, which contains all the transactions that could be mined in the future. Hardhat Network's mempool follows the same rules as geth. This means, among other things, that:

> - Transactions with a higher gas price are included first

Reviewer: ```suggestion
- Transactions with a higher gas/fees prices are included first
```

> - If two transactions can be included and both have the same gas price, the one that was received first is included first

Reviewer: See: https://github.com/nomiclabs/hardhat/pull/1657/files#r674852709

Author: Addressed in https://github.com/nomiclabs/hardhat/pull/1657/commits/20aa02e0a31da82939a609c8ccb57c6d38b9401d

- If a transaction is invalid (for example, its nonce is lower than the nonce of the address that sent it), the transaction is dropped.

You can get the list of pending transactions that will be included in the next block by using the "pending" block tag:

```js
const pendingBlock = await network.provider.send("eth_getBlockByNumber", [
  "pending",
  false,
]);
```

### Removing and replacing transactions

Transactions in the mempool can be removed using the `hardhat_dropTransaction` method:

```js
const txHash = "0xabc...";
await network.provider.send("hardhat_dropTransaction", [txHash]);
```

> You can also replace a transaction by sending a new one with the same nonce as the one that it's already in the mempool but with a higher gas price. Keep in mind that, like in Geth, for this to work the new gas price has to be at least 10% higher than the gas price of the current transaction.

Reviewer: ```suggestion
You can also replace a transaction by sending a new one with the same nonce as the one that it's already in the mempool but with a higher gas price. Keep in mind that, like in Geth, for this to work the new gas/fees prices has to be at least 10% higher than the gas price of the current transaction.
```

Author: Addressed in https://github.com/nomiclabs/hardhat/pull/1657/commits/5dee7c7aceb2857be602261237b276c993f50dc9

#### `eth_chainId`

#### `eth_coinbase`

#### `eth_estimateGas`

#### `eth_gasPrice`

#### `eth_getBalance`

#### `eth_getBlockByHash`

#### `eth_getBlockByNumber`

#### `eth_getBlockTransactionCountByHash`

#### `eth_getBlockTransactionCountByNumber`

#### `eth_getCode`

#### `eth_getFilterChanges`

#### `eth_getFilterLogs`

#### `eth_getLogs`

#### `eth_getStorageAt`

#### `eth_getTransactionByBlockHashAndIndex`

#### `eth_getTransactionByBlockNumberAndIndex`

#### `eth_getTransactionByHash`

#### `eth_getTransactionCount`

#### `eth_getTransactionReceipt`

#### `eth_mining`

#### `eth_newBlockFilter`

#### `eth_newFilter`

#### `eth_newPendingTransactionFilter`

#### `eth_pendingTransactions`

#### `eth_sendRawTransaction`

#### `eth_sendTransaction`

#### `eth_sign`

#### `eth_signTypedData_v4`

#### `eth_subscribe`

#### `eth_syncing`

#### `eth_uninstallFilter`

#### `eth_unsubscribe`

#### `net_listening`

#### `net_peerCount`

#### `net_version`

#### `web3_clientVersion`

#### `web3_sha3`

### Hardhat network methods

#### `hardhat_addCompilationResult`

Add information about compiled contracts

#### `hardhat_dropTransaction`

Remove a transaction from the mempool

> #### `hardhat_getStackTraceFailuresCount`

Reviewer: I think I wouldn't document this. It's mostly an internal API.

Author: Addressed in https://github.com/nomiclabs/hardhat/pull/1657/commits/368a1ff0a7678b23280d5654d9dc67b3fc2d1674

#### `hardhat_impersonateAccount`

> see the [Mainnet Forking guide](../../guides/mainnet-forking.md)

Reviewer: ```suggestion
See the [Mainnet Forking guide](../../guides/mainnet-forking.md)
```

Reviewer: Note: This also works when you are not forking a remote network.

> #### `hardhat_intervalMine`

Reviewer: Same. Also internal.

Author: Addressed in https://github.com/nomiclabs/hardhat/pull/1657/commits/fc6bc587962d4c7c56e76e7ebc2b0043400bc67c

#### `hardhat_reset`

> see the [Mainnet Forking guide](../../guides/mainnet-forking.md)

Reviewer: ```suggestion
See the [Mainnet Forking guide](../../guides/mainnet-forking.md)
```

#### `hardhat_setBalance`

Modifies the balance of an account.

#### `hardhat_setCode`

Modifies the code of an account.

#### `hardhat_setLoggingEnabled`

Enable or disable logging in Hardhat Network

#### `hardhat_setMinGasPrice`

> change the minimum gas price accepted by the network (in wei)

Reviewer: ```suggestion
Change the minimum gas price accepted by the network (in wei)
```

Reviewer: This method doesn't work when running London

#### `hardhat_setNonce`

> Modifies an account's nonce by overwriting it. Throws an InvalidInputError if nonce is smaller than the current one. The reason for this restriction is to avoid collisions when deploying contracts using the same nonce more than once.

Reviewer: ```suggestion
Modifies an account's nonce by overwriting it. Throws an `InvalidInputError` if nonce is smaller than the current one. The reason for this restriction is to avoid collisions when deploying contracts using the same nonce more than once.
```

#### `hardhat_setStorageAt`

Writes a single position of an account's storage. The storage position index must not exceed 2^256, and the value to write must be exactly 32 bytes long.

#### `hardhat_stopImpersonatingAccount`

> see the [Mainnet Forking guide](../../guides/mainnet-forking.md)

Reviewer: ```suggestion
See the [Mainnet Forking guide](../../guides/mainnet-forking.md)
```

> ### Special testing/debugging methods

Reviewer: I consider `debug_traceTransaction` a standard method.

Author: Addressed in https://github.com/nomiclabs/hardhat/pull/1657/commits/615def3a3ef4e44c89549512b2fc4f916d74fcdb

#### `debug_traceTransaction`

Get debug traces of already-mined transactions.

To get a trace, call this method with the hash of the transaction as its argument:

```js
const trace = await hre.network.provider.send("debug_traceTransaction", [
  "0x123...",
]);
```

You can also selectively disable some properties in the list of steps:

```js
const trace = await hre.network.provider.send("debug_traceTransaction", [
  "0x123...",
  {
    disableMemory: true,
    disableStack: true,
    disableStorage: true,
  },
]);
```

##### Known limitations

- You can't trace transactions that use a hardfork older than [Spurious Dragon](https://ethereum.org/en/history/#spurious-dragon)
- The last step of a message is not guaranteed to have a correct value in the `gasCost` property

#### `evm_increaseTime`

> same as Ganache.

Reviewer: ```suggestion
Same as Ganache.
```

#### `evm_mine`

> same as Ganache

Reviewer: ```suggestion
Same as Ganache
```

#### `evm_revert`

> same as Ganache.

Reviewer: ```suggestion
Same as Ganache.
```

#### `evm_setAutomine`

Enables or disables, based on the single boolean argument, the automatic mining of new blocks with each new transaction submitted to the network.

#### `evm_setBlockGasLimit`

#### `evm_setIntervalMining`

Enables (with a numeric argument greater than 0) or disables (with a numeric argument equal to 0), the automatic mining of blocks at a regular interval of milliseconds, each of which will include all pending transactions.

#### `evm_setNextBlockTimestamp`

This method works like `evm_increaseTime`, but takes the exact timestamp that you want in the next block, and increases the time accordingly.

#### `evm_snapshot`

> same as Ganache.

Reviewer: ```suggestion
Same as Ganache.
```
